# Pandas Homework - Heroes of Pymoli

In [1]:
#Import Dependencies
import pandas as pd

In [2]:
#File to Load
file_data = "Resources/purchase_data.csv"

#Read the modified csv file and store into Pandas DataFrame
file_data_df = pd.read_csv(file_data, encoding="utf-8")

In [3]:
#Display the total number of players
total_count = len(file_data_df["SN"].unique())
total_count_df = pd.DataFrame({"Total Players":[total_count]})
total_count_df

Total Players
0            576

In [4]:
#Run basic calculations to obtain number of unique items.
unique_item = len(file_data_df["Item Name"].unique())
average_price = file_data_df["Price"].mean()
number_purchases = len(file_data_df["Purchase ID"])
total_revenue = file_data_df["Price"].sum()
summary_df = pd.DataFrame({"Number of Unique Items":[unique_item],
                           "Average Price":[average_price],
                           "Number of purchases": number_purchases,
                           "Total Revenue": total_revenue})
#Format to $
summary_df["Average Price"] = summary_df["Average Price"].map("${:.2f}".format)
summary_df["Total Revenue"] = summary_df["Total Revenue"].map("${:,.2f}".format)
#Display the summary Data Frame
summary_df

Number of Unique Items Average Price  Number of purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

In [5]:
#Using GroupBy
player_count = file_data_df.groupby(['Gender'])['SN'].nunique()
total_gen_players = file_data_df['SN'].nunique()
#Finding the percentage 
player_percent = player_count/total_gen_players * 100
#Converting into a Dataframe
gender_demo_df = pd.DataFrame({"Total Count":player_count, "Percentage of Players":player_percent})
#Format to %
gender_demo_df["Percentage of Players"] = gender_demo_df["Percentage of Players"].map("{0:,.2f}%".format)
gender_demo_df.index.name = None
#Sort Values
gender_demographics_df = gender_demo_df.sort_values("Total Count", ascending = False)
#Display Values
gender_demographics_df 

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%

In [6]:
#Using GroupBy in order to separate the data into fields according to "gender" values
grouped_gender_df = file_data_df.groupby(['Gender'])
gender_purchase_count = grouped_gender_df['Purchase ID'].count()
#Average purchase price
gender_avg_price = round(grouped_gender_df['Price'].mean(),2)
#Total Purchase value
gender_total_purchase = grouped_gender_df['Price'].sum()
#Average Total per person
gender_avg_perperson = gender_total_purchase/player_count

In [7]:
#Converting a GroupBy object into a DataFrame
p_gender_df = pd.DataFrame({"Purchase Count": gender_purchase_count,
                            "Average Purchase Price": gender_avg_price,
                            "Total Purchase Value": gender_total_purchase,
                            "Avg Total Purchase per Person":gender_avg_perperson})
#Format to $
p_gender_df["Average Purchase Price"] = p_gender_df["Average Purchase Price"].map("${:.2f}".format)
p_gender_df["Total Purchase Value"] = p_gender_df["Total Purchase Value"].map("${:,.2f}".format)
p_gender_df["Avg Total Purchase per Person"] =  p_gender_df["Avg Total Purchase per Person"].map("${:.2f}".format)
#Display DataFrame
p_gender_df

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

In [8]:
#Establish bins for ages
bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]

#Create the names for the bins
group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

In [9]:
#Age Demographics 
file_data_df["Total Count"] = pd.cut(file_data_df["Age"], bins, labels=group_names, include_lowest=True)
#Using Groupby
new_df = file_data_df.groupby(['Total Count'])
count_age = new_df["SN"].nunique()
count_age_total = file_data_df["SN"].nunique()
age_percent = count_age / count_age_total * 100
#Convert into DataFrame
age_demog_df = pd.DataFrame({"Total Count":count_age, "Percentage of Players":age_percent})
#Format to %
age_demog_df["Percentage of Players"] = age_demog_df["Percentage of Players"].map("{0:,.2f}%".format)
age_demog_df.index.name = None
#Display DataFrame
age_demog_df

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

In [10]:
#Purchasing Analysis (Age)
#Functions
purchase_a_df = new_df["Purchase ID"].count()
mean_a_df = new_df["Price"].mean()
price_a_df = new_df["Price"].sum()
avg_a_df = price_a_df / count_age
#Covert into Data Frame
p_age_df = pd.DataFrame({"Purchase Count":purchase_a_df, "Average Purchase Price":mean_a_df,
                         "Total Purchase Value":price_a_df,"Avg Total Purchase per Person":avg_a_df})
# Format to $
p_age_df["Average Purchase Price"] = p_age_df["Average Purchase Price"].map("${:.2f}".format)
p_age_df["Total Purchase Value"] = p_age_df["Total Purchase Value"].map("${:,.2f}".format)
p_age_df["Avg Total Purchase per Person"] = p_age_df["Avg Total Purchase per Person"].map("${:.2f}".format)
# Name index 
p_age_df.index.name = 'Age Ranges'
# Display DataFrame
p_age_df

Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05            $1,114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

In [11]:
#Top Spenders
spenders_df = file_data_df.groupby(["SN"])["Purchase ID"].count()
average_s_df = file_data_df.groupby(["SN"])["Price"].mean()
total_p_df = spenders_df * average_s_df
#Convert into DataFrame
top_s_df = pd.DataFrame({"Purchase Count":spenders_df, "Average Purchase Price":average_s_df,
                         "Total Purchase Value":total_p_df})
#Sort Values
sort_top_df = top_s_df.sort_values("Total Purchase Value", ascending = False)
#Format to $
sort_top_df["Average Purchase Price"] = sort_top_df["Average Purchase Price"].map("${:.2f}".format)
sort_top_df["Total Purchase Value"] = sort_top_df["Total Purchase Value"].map("${:.2f}".format)
#Display DataFrame
sort_top_df.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

In [12]:
#Most Popular Items
popular_df = file_data_df.groupby(["Item ID", "Item Name"])["Purchase ID"].count()
total_item_p = file_data_df.groupby(["Item ID", "Item Name"])["Price"].sum()
item_price = total_item_p/popular_df

#Convert into DataFrame
most_item_df = pd.DataFrame({"Purchase Count":popular_df, "Item Price": item_price, 
                             "Total Purchase Value": total_item_p})
#Sort Values
sort_item_df = most_item_df.sort_values("Purchase Count", ascending = False)

#Format to $
sort_item_df["Item Price"] = sort_item_df["Item Price"].map("${:.2f}".format)
sort_item_df["Total Purchase Value"] = sort_item_df["Total Purchase Value"].map("${:.2f}".format)

#Display DataFrame
sort_item_df.head()


Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

In [13]:
#Most Profitable Items
most_prof_df = most_item_df.sort_values("Total Purchase Value", ascending = False)

#Format to $
most_prof_df["Item Price"] = most_prof_df["Item Price"].map("${:.2f}".format)
most_prof_df["Total Purchase Value"] = most_prof_df["Total Purchase Value"].map("${:.2f}".format)

#Display DataFrame
most_prof_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

# Three observable trends based on the data

1. Aproximately 84% of Pymoli's players are male and their total purchase per person is on average \\$ 4.07, that is 40 cents less than what women spend. Since there are more male players than women, the male's total purchase value is equal to \\$ 1,967.64 making it the highest value among genders, with the second highest being only $ 361.94 for women.


2. There are 576 players in Pymoli, aproximately 44.8% of the players are between the ages of 20 to 24. This age group is the one that gives the company the most profit, with their total purchase value being $ 1,114.06. 
However, the age group that is between 35 to 39 years old is the one that on average spends more per person. 


3. The top two items that are most profitable for Pymoli are the Final Critic and the Oathbreaker,Last Hope of the breaking Storm. This two items have been bought thirteen and twelve times respectively, making them the most popular items in the game. 

